### PR TITLE
VOOD-169: General refactoring.

### PR DESCRIPTION
**Pull request by GinoCardillo-OST**
* Fix a bug where references to null had the incorrect id
* Refactor PanelViewInputObjectDiagramReader to reduce complexity
* Remove Java specific parts in PanelViewInputObjectDiagramReader

### CODE REVIEW

File: src/object-diagram/logic/test-suites/miscellaneous/tests/full/writer/expected.puml [hunk 1]
```diff
@@ -1,6 +1,8 @@
 @startuml
 map "<<[StackFrame]>>" as _0f07b6f2a31cbd2e26428c51e8660b0dffe6e352 {
   borderWidth => 3 (int)
+  topLeft => null
+  pNull => null
```
Comment (GinoCardillo-OST) [inline src/object-diagram/logic/test-suites/miscellaneous/tests/full/writer/expected.puml:5]: Another bug fixed here. Variables with a null reference were missing.

File: src/object-diagram/logic/reader/panelViewInputObjectDiagramReader.ts [hunk 2]
```diff
@@ -1,15 +1,23 @@
...
-import { createStructureId } from '../../model/structureId';
+import { createStructureId, StructureId } from '../../model/structureId';
 import { ObjectDiagramReader } from './objectDiagramReader';
 
-const isUndefinedOrPrimitiveJavaType = (type: string | undefined): type is undefined | PrimitiveJavaType =>
-  type === undefined || isPrimitiveJavaType(type);
+const isObject = (variable: PanelViewVariable): boolean => hasObjectPrefix(variable.id);
+
+const isNullOfObject = (variable: PanelViewVariable): boolean =>
+  hasNullPrefix(variable.id) && variable.incomingRelations !== undefined && Boolean(variable.incomingRelations[0].relationName);
```
Comment (alexandre-lagadec-ost) [inline src/object-diagram/logic/reader/panelViewInputObjectDiagramReader.ts:14]: Does this function test for `null` references coming from an object?

File: src/object-diagram/logic/reader/panelViewInputObjectDiagramReader.ts [hunk 3]
```diff
@@ -1,15 +1,23 @@
...
+
+const isNullOfVariable = (variable: PanelViewVariable): boolean =>
+  hasNullPrefix(variable.id) && variable.incomingRelations !== undefined && !variable.incomingRelations[0].relationName;
```
Comment (alexandre-lagadec-ost) [inline src/object-diagram/logic/reader/panelViewInputObjectDiagramReader.ts:17]: Does this function test for `null` references coming from a stack frame?

**Review by PascalSchuermannOst — APPROVED** [review]

File: src/object-diagram/logic/test-suites/linked-list/test-suite.json [hunk 4]
```diff
@@ -2,5 +2,5 @@
   "ignore": false,
   "only": false,
   "title": "linked list",
-  "_source": "https://github.com/hediet/vscode-debug-visualizer/blob/master/demos/java/src/app/App.java"
+  "_source": "demos/java/src/DoublyLinkedListDemo.java"
```
[the remark this answers is not in the document]
Comment (GinoCardillo-OST) [reply src/object-diagram/logic/test-suites/linked-list/test-suite.json:5]: Then why do you want to change it back? The test uses **our** code.

File: src/object-diagram/logic/reader/panelViewInputObjectDiagramReader.ts (hunk 2 above)
Comment (GinoCardillo-OST) [reply src/object-diagram/logic/reader/panelViewInputObjectDiagramReader.ts:14]: Yes.

File: src/object-diagram/logic/reader/panelViewInputObjectDiagramReader.ts (hunk 3 above)
Comment (GinoCardillo-OST) [reply src/object-diagram/logic/reader/panelViewInputObjectDiagramReader.ts:17]: Coming from a variable

File: src/object-diagram/logic/reader/panelViewInputObjectDiagramReader.ts [hunk 5]
```diff
@@ -22,88 +30,100 @@ export class PanelViewInputObjectDiagramReader implements ObjectDiagramReader<Pa
...
+    }
+    return [
+      {
+        relationName: variable.name,
+        parentId: stackFrameId,
+      },
+    ];
+  }
+
+  private createFieldsFromRelations(relations: VariableRelation[], variable: PanelViewVariable): Field[] {
```
[the remark this answers is not in the document]
Comment (GinoCardillo-OST) [reply src/object-diagram/logic/reader/panelViewInputObjectDiagramReader.ts:91]: As the name `getRelationsOfVariable` implies, the variable is the center piece and thus is the first parameter.
As the name `createFieldsFromRelations` implies, the relations are the center pieces and thus are the first parameter.

File: src/object-diagram/logic/test-suites/linked-list/test-suite.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexandre-lagadec-ost) [reply src/object-diagram/logic/test-suites/linked-list/test-suite.json:5]: Sorry, I confused what was added with what was removed. You can safely ignore my suggestion.

File: src/object-diagram/logic/reader/panelViewInputObjectDiagramReader.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexandre-lagadec-ost) [reply src/object-diagram/logic/reader/panelViewInputObjectDiagramReader.ts:91]: Good point(s).

File: src/object-diagram/logic/reader/panelViewInputObjectDiagramReader.ts [hunk 6]
```diff
@@ -22,88 +30,100 @@ export class PanelViewInputObjectDiagramReader implements ObjectDiagramReader<Pa
...
-      primitiveValues?.forEach((primitiveValue) => {
+    for (const variable of variables.values()) {
+      const relations = this.getRelationsOfVariable(variable, stackFrameId);
+      if (isObject(variable)) {
+        const structureId = createStructureId(variable.id);
+        structures.push(this.createStructure(variable, structureId));
+        fields.push(...this.createFieldsFromPrimitiveValues(variable.primitiveValues, structureId));
+        references.push(...this.createReferencesFromRelations(relations, stackFrameId, structureId, variables));
+      } else if (isPrimitiveVariable(variable) || isNullOfObject(variable)) {
+        fields.push(...this.createFieldsFromRelations(relations, variable));
```
[the remark this answers is not in the document]
Comment (alexandre-lagadec-ost) [reply src/object-diagram/logic/reader/panelViewInputObjectDiagramReader.ts:41]: No longer relevant (see https://github.com/GinoCardillo-OST/Visual-OO-Debugger/pull/39#discussion_r872958197) -> Resolved

**Review by alexandre-lagadec-ost — APPROVED** [review]

**Pull request merged**